File: frontend/src/utils/notificationDestination.test.js
import { describe, it, expect } from 'vitest';
import { notificationDestination } from './notificationDestination.js';

describe('notification destinations', () => {
  it('opens friends for requests and acceptance even with a legacy dashboard URL', () => {
    for (const type of ['friend_request', 'friend_accepted']) {
      expect(notificationDestination({ type, url: '/dashboard' })).toBe('/friends');
    }
  });
  it('opens the related group and encodes identifiers', () => {
    expect(notificationDestination({ groupId: 'trip 1' })).toBe('/groups/trip%201');
    expect(notificationDestination({ url: '/groups/trip1' })).toBe('/groups/trip1');
  });
  it('rejects external, script and unknown destinations', () => {
    for (const url of [
      'https://example.com/friends',
      '//evil.test/groups/a',
      'javascript:alert(1)',
      '/admin',
    ]) {
      expect(notificationDestination({ url })).toBe('/dashboard');
    }
  });
});

File: native-android/app/src/test/java/com/paymatrix/app/data/NotificationDestinationTest.kt
package com.paymatrix.app.data

import org.junit.Assert.assertEquals
import org.junit.Test

class NotificationDestinationTest {
    @Test fun friendEventsOverrideLegacyDashboardLink() {
        listOf("friend_request", "friend_accepted").forEach {
            assertEquals("friends", NotificationDestination.resolve(type = it, url = "/dashboard"))
        }
    }
    @Test fun groupEventsAndWebLinksOpenNativeGroupRoute() {
        assertEquals("group/trip_1", NotificationDestination.resolve(groupId = "trip_1"))
        assertEquals("group/trip_1", NotificationDestination.resolve(url = "/groups/trip_1"))
        assertEquals("group/trip_1", NotificationDestination.resolve(url = "group/trip_1"))
        assertEquals("friends", NotificationDestination.resolve(url = "https://paymatrixapp.online/friends"))
        assertEquals("friends", NotificationDestination.resolve(url = "https://pay-matrix.vercel.app/friends"))
    }
    @Test fun widgetActionsRetainGroupContext() {
        assertEquals("expense/trip_1", NotificationDestination.resolve(url = "expense/trip_1"))
        assertEquals("scanner?groupId=trip_1", NotificationDestination.resolve(url = "scanner?groupId=trip_1"))
        assertEquals("scanner", NotificationDestination.resolve(url = "scanner?groupId=../private"))
    }
    @Test fun unsafeAndUnknownLinksFallBack() {
        listOf("https://evil.test/friends", "//evil.test/friends", "javascript:alert(1)", "/admin", "/groups/a/b").forEach {
            assertEquals("dashboard", NotificationDestination.resolve(url = it))
        }
    }
}
